namui-cli/src/build/build_server/code_watcher.rs

```diff
@@ -0,0 +1,164 @@
+use crate::debug_println;
+use cargo_metadata::MetadataCommand;
+use futures::{executor::block_on, lock::Mutex};
+use notify::{DebouncedEvent, INotifyWatcher, RecursiveMode, Watcher};
+use regex::Regex;
+use std::{collections::HashSet, path::PathBuf, sync::Arc, thread, time::Duration};
+use tokio::sync::oneshot::{self, Receiver, Sender};
+
+struct CodeWatcherContext {
+    was_changed: bool,
+    sender: Option<Sender<()>>,
+}
+
+enum CodeWatcherChangeState {
+    Changed,
+    NotChanged(Receiver<()>),
+}
+
+pub struct CodeWatcher {
+    watcher: INotifyWatcher,
+    context: Arc<Mutex<CodeWatcherContext>>,
+    watching_paths: HashSet<String>,
+    manifest_path: String,
+}
+
+impl CodeWatcher {
+    pub fn new(manifest_path: String) -> Self {
+        let context = Arc::new(Mutex::new(CodeWatcherContext {
+            was_changed: true,
+            sender: None,
+        }));
+        let mut root_watching_path = PathBuf::from(&manifest_path);
+        root_watching_path.pop();
+        let root_watching_path = root_watching_path.join("src").to_string_lossy().to_string();
+
+        let (watcher_sender, watcher_receiver) = std::sync::mpsc::channel::<DebouncedEvent>();
+        let mut watcher = notify::watcher(watcher_sender, Duration::from_secs(1)).unwrap();
+        if let Err(error) = watcher.watch(root_watching_path, RecursiveMode::Recursive) {
+            panic!("watch failed {:?}", error);
+        }
+
+        let context_clone = context.clone();
+        thread::spawn(move || -> ! {
+            loop {
+                match watcher_receiver.recv() {
+                    Ok(event) => {
+                        match event {
+                            DebouncedEvent::Create(_)
+                            | DebouncedEvent::Remove(_)
+                            | DebouncedEvent::Rename(_, _)
+                            | DebouncedEvent::Write(_) => {
+                                debug_println!("CodeWatcher: locking code_watcher.context...");
+                                let mut context = block_on(context_clone.lock());
+                                debug_println!("CodeWatcher: code_watcher.context locked");
+
+                                match context.sender.take() {
+                                    Some(sender) => {
+                                        if let Err(_) = sender.send(()) {
+                                            panic!("send failed in watcher")
+                                        }
+                                    }
+                                    None => {
+                                        context.was_changed = true;
```

Comment:
If there are 2 DebouncedEvent, and if `context.sender` is not None, this thread will take `context.sender` and send it, and will make context.was_changed true, so it build twice, isn't it?

Reply:
I think code `while watcher_receiver.try_recv().is_ok() {}` can flush events in mpsc.